src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Cache/CacheFallbackOperations.cs

```diff
@@ -316,7 +316,7 @@ internal class CacheFallbackOperations
                             foreach (MsalAccountCacheItem accountCacheItem in accounts)
                             {
                                 if (rtCacheItem.HomeAccountId.Equals(accountCacheItem.HomeAccountId, StringComparison.OrdinalIgnoreCase)
-                                    && accountCacheItem.PreferredUsername.Equals(upn, StringComparison.OrdinalIgnoreCase))
+                                    && (string.IsNullOrEmpty(upn) || accountCacheItem.PreferredUsername.Equals(upn, StringComparison.OrdinalIgnoreCase)))
```

Comment:
>(string.IsNullOrEmpty(upn)  [](start = 39, length = 27)

is this the fix we allowing me to share tokens between MSAL v3 and ADAL v5?

Reply:
this fixes the case where the user calls AcquireTokenSilentAsync() with no account/user information.  This check is already done in the adal cache filtering but is ignored in the msal fallback case so any msal deserialization won't find the token here if the user/account information is not specified.

---
In reply to: [260969491](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#discussion_r260969491) [](ancestors = 260969491)